### PR TITLE
Add support for QUEST_OBJECTIVE_CRITERIA_TREE

### PR DESCRIPTION
**Pull request by rstebler**
**Changes proposed:**

-  Add support for QUEST_OBJECTIVE_CRITERIA_TREE

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Tests performed:** (Does it build, tested in-game, etc.)

- [x] Does it build?
- [x] Tested in-game?

Adds a QuestObjectiveCriteriaMgr that handles the criteria for quest objectives of type QUEST_OBJECTIVE_CRITERIA_TREE.
Progress is stored in new tables character_queststatus_objectives_criteria and character_queststatus_objectives_criteria_progress.

Tested with quests 38612, 42688, 43562, 45422

### CODE REVIEW

File: src/server/game/Achievements/CriteriaHandler.cpp [hunk 1]
```diff
@@ -2305,6 +2320,8 @@ void CriteriaMgr::LoadCriteriaList()
             }
             else if (tree->ScenarioStep)
                 criteria->FlagsCu |= CRITERIA_FLAG_CU_SCENARIO;
+            else if (tree->QuestObjective)
+                criteria->FlagsCu |= CRITERIA_FLAG_CU_PLAYER;
```
Comment (Shauren) [inline src/server/game/Achievements/CriteriaHandler.cpp]: Why are you reusing PLAYER flag here instead of adding another flag? This forces you to make unrelated changes to `PlayerAchievementMgr`

File: src/server/game/Quests/QuestObjectiveCriteriaMgr.cpp [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        criteriaUpdate.Flags = 0;
+
+        criteriaUpdate.CurrentTime = itr->second.Date;
+        criteriaUpdate.CreationTime = 0;
+
+        SendPacket(criteriaUpdate.Write());
+    }
+}
+
+void QuestObjectiveCriteriaMgr::CompletedObjective(QuestObjective const* objective, Player* referencePlayer)
```
Comment (tkrokli) [inline src/server/game/Quests/QuestObjectiveCriteriaMgr.cpp]:
```cpp
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Quests/QuestObjectiveCriteriaMgr.cpp:226:93:
 fatal error: unused parameter 'referencePlayer' [-Wunused-parameter]

void QuestObjectiveCriteriaMgr::CompletedObjective(QuestObjective const* objective, Player* referencePlayer)
                                                                                            ^
```

File: src/server/game/Achievements/CriteriaHandler.cpp (hunk 1 above)
Comment (rstebler) [inline src/server/game/Achievements/CriteriaHandler.cpp]: You are right, thanks. Using CRITERIA_FLAG_CU_QUESTOBJECTIVE now.

File: src/server/game/Quests/QuestObjectiveCriteriaMgr.cpp (hunk 2 above)
Comment (rstebler) [inline src/server/game/Quests/QuestObjectiveCriteriaMgr.cpp]: Thanks, corrected.

File: sql/base/characters_database.sql [hunk 3]
```diff
@@ -1186,6 +1186,55 @@ LOCK TABLES `character_queststatus_objectives` WRITE;
...
 
 --
+-- Table structure for table `character_queststatus_objectives_criteria`
+--
+
+DROP TABLE IF EXISTS `character_queststatus_objectives_criteria`;
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!40101 SET character_set_client = utf8 */;
+CREATE TABLE `character_queststatus_objectives_criteria` (
+  `guid` int(10) unsigned NOT NULL,
```
Comment (joschiwald) [inline sql/base/characters_database.sql]: guid is `bigint(20) unsigned`

Comment (rstebler) [inline sql/base/characters_database.sql]: Thanks, corrected.

File: src/server/game/Achievements/CriteriaHandler.h [hunk 4]
```diff
@@ -395,7 +402,8 @@ class TC_GAME_API CriteriaMgr
     CriteriaList _criteriasByType[CRITERIA_TYPE_TOTAL];
     CriteriaList _guildCriteriasByType[CRITERIA_TYPE_TOTAL];
     CriteriaList _scenarioCriteriasByType[CRITERIA_TYPE_TOTAL];
-
+    CriteriaList _questObjectiveCriteriasByType[CRITERIA_TYPE_TOTAL];
+
```
Comment (joschiwald) [inline src/server/game/Achievements/CriteriaHandler.h]: whitespace
